### ISSUE TITLE
make the relationship time-dependant

### ISSUE THREAD

**Issue by valentinomariotto** [staudenmeir/laravel-adjacency-list#144]
Hi, I really appreciate your work with this package, saved me a lot of headache. 

I have a simple 1 parent to many children relationship.
The main model is my User model, but the relationship data is stored in another table.
I have simple parent and children function directly in the User model, and I use your functions to get all the descendants or the ancestors from the relationships table, and then look up the resulting ids in the Users table. Not ideal but good enough.

Now, I have to make these relationships time dependent. Meaning that there might be many parents for each user, but I'll be normally considering only the LATEST record, except when a date is provided.
I rebuilt the `parent(Carbon $date=null)` and `children(Carbon $date=null)` to accept a date as an optional argument and return the parent or children at that point in time, and use the latest records otherwise.

```
public function parent(Carbon $date=null)
    {
        return $this->belongsToMany(self::class, 'user_relationships', 'user_id', 'sponsor_id')
                    ->wherePivotBetween('created_at', [Carbon::parse('2000-01-01'), (@$date ?: Carbon::now())])
                    ->orderByPivot('created_at', 'desc')
                    ->limit(1);
    }
```

Now, is there any way to achieve the same result with this package?
I suppose I should use the graph mode and `withRecursiveQueryConstraint`.. But have you already come across this problem? Have you got any suggestions?

**Comment by staudenmeir** [comment]
Hi @valentinomariotto,
Please share the migrations of both tables.

> I suppose I should use the graph mode

Did you already [get started](https://github.com/staudenmeir/laravel-adjacency-list#graphs-getting-started) with that and try the "normal" `descendants`/`ancestors` relationships?

**Comment by valentinomariotto** [comment]
well, not yet.. Because I realize that I need an additional table (edges) for using the graph mode and it seems too much overhead and complexity for my needs.

These are the migrations:

Schema::create('users', function (Blueprint $table) {
            $table->bigIncrements('id');
            $table->string('username', 32)->unique();
            $table->string('email')->unique();
            $table->string('password');
            $table->rememberToken();
            $table->timestamps();
            $table->softDeletes();
        });

Schema::create('user_relationships', function (Blueprint $table) {
            $table->bigInteger('user_id')->unsigned();
            $table->bigInteger('sponsor_id')->unsigned();
            $table->integer('matrix_level')->unsigned();
            $table->timestamps();
            $table->timestamp('superseeded_at')->nullable();

            $table->primary(['user_id', 'sponsor_id']);
        });

**Comment by staudenmeir** [comment]
`user_relationships` is the additional table ;-) `user_id` and `sponsor_id` are the keys you need to set up in your model as described in the README.

**Comment by valentinomariotto** [comment]
ooooh, now I understand.

But do I really need the graph mode? For now, I tried with this constraint on the tree mode. I returns results, but the constraints don't apply

```
$descendants_ids = UserRelationships::withRecursiveQueryConstraint(function ($q) use ($date) {
                    $q->whereDate('user_relationships.created_at', '<', (@$date ?: Carbon::now()) )
                        ->where(function ($qq) use ($date) {
                            $qq->whereNull('user_relationships.superseeded_at')
                                ->orWhereDate('user_relationships.superseeded_at', '>', (@$date ?: Carbon::now()) );
                        });
                }, function () use ($user_id) {
                    return UserRelationships::firstWhere('user_id', $user_id)
                                            ->children() 
                                            //or ->descendants()
                                            ->pluck('user_id');
                });
```

**Comment by staudenmeir** [comment]
Can you log this query with `->children()` and `->descendants()`?

**Comment by valentinomariotto** [comment]
Hi, my apologies for the delay.
Let's consider an example.
The relationships are as follows:
```
|user_id|sponsor_id|matrix_level|created_at         |updated_at         |superseded_at     |
|-------|----------|------------|-------------------|-------------------|-------------------|
|24     |1         |1           |2022-01-31 17:08:18|2022-11-24 09:21:13|2037-12-31 00:00:00|
|25     |24        |2           |2022-01-31 17:32:32|2022-11-24 09:21:13|2037-12-31 00:00:00|
|234    |24        |3           |2022-04-13 13:49:50|2022-11-24 09:21:13|2022-10-21 12:00:00|
|234    |25        |3           |2022-10-21 12:00:00|2022-11-24 09:21:13|2037-12-31 00:00:00|
|437    |234       |4           |2022-06-23 16:24:12|2022-11-24 09:21:14|2037-12-31 00:00:00|
```
the relationship between 234 and 24 was superseded by a new one between 234 and 25. 

I set up a test route, here with the `ancestors()` method:
```
Route::get('/testrelationships', function () {

        Illuminate\Support\Facades\DB::connection()->enableQueryLog();

        $date = Carbon\Carbon::parse('2022-11-03 12:00:00');
        $upline_ids = \App\Models\UserRelationships::withRecursiveQueryConstraint(function ($q) use ($date) {
                $q->whereDate('user_relationships.created_at', '<=', $date  )
                    ->where(function ($qq) use ($date) {
                        $qq->whereNull('user_relationships.superseded_at')
                            ->orWhereDate('user_relationships.superseded_at', '>', $date  );
                    });
            }, function () {
                return \App\Models\UserRelationships::where('user_id', 437)
                                        ->latest()
                                        ->first()
                                        ->**ancestors()**
                                        ->pluck('user_id')
                                        ->toArray();
            });
        echo implode('<br>', $upline_ids );

        $queries = Illuminate\Support\Facades\DB::getQueryLog();
        dd ( end( $queries ));

    });
```

these are the results: 

```
234
234
24
25
24
array:3 [▼ // routes/web.php:67
  "query" => "with recursive `laravel_cte` as ((select *, -1 as `depth`, cast(`user_id` as char(65535)) as `path` from `user_relationships` where `user_relationships`.`user_id` = ?) union all (select `user_relationships`.*, `depth` - 1 as `depth`, concat(`path`, ?, `user_relationships`.`user_id`) from `user_relationships` inner join `laravel_cte` on `laravel_cte`.`sponsor_id` = `user_relationships`.`user_id` where date(`user_relationships`.`created_at`) <= ? and (`user_relationships`.`superseeded_at` is null or date(`user_relationships`.`superseeded_at`) > ?))) select `user_id` from `laravel_cte` ◀"
  "bindings" => array:4 [▼
    0 => 234
    1 => "."
    2 => "2022-11-03"
    3 => "2022-11-03"
  ]
  "time" => 8.08
]
```


using the **children()** method (not changing anything else in the previous code):
```
501
502
507
511
array:3 [▼ // routes/web.php:67
  "query" => "select `user_id` from `user_relationships` where `user_relationships`.`sponsor_id` = ? and `user_relationships`.`sponsor_id` is not null"
  "bindings" => array:1 [▼
    0 => 437
  ]
  "time" => 4.94
]
```
the constraint are ignored.


using the **descendants()** method (not changing anything else in the previous code):
```
501
502
507
511
array:3 [▼ // routes/web.php:67
  "query" => "with recursive `laravel_cte` as ((select *, 1 as `depth`, cast(`user_id` as char(65535)) as `path` from `user_relationships` where `user_relationships`.`sponsor_id` = ? and `user_relationships`.`sponsor_id` is not null) union all (select `user_relationships`.*, `depth` + 1 as `depth`, concat(`path`, ?, `user_relationships`.`user_id`) from `user_relationships` inner join `laravel_cte` on `laravel_cte`.`user_id` = `user_relationships`.`sponsor_id` where date(`user_relationships`.`created_at`) <= ? and (`user_relationships`.`superseded_at` is null or date(`user_relationships`.`superseded_at`) > ?))) select `user_id` from `laravel_cte` ◀"
  "bindings" => array:4 [▼
    0 => 437
    1 => "."
    2 => "2022-11-03"
    3 => "2022-11-03"
  ]
  "time" => 8.28
]
```

**Comment by valentinomariotto** [comment]
if I apply the constraint on the results too:
```
Route::get('/testrelationships', function () {

        Illuminate\Support\Facades\DB::connection()->enableQueryLog();

        $date = Carbon\Carbon::parse('2022-11-03 12:00:00');
        $upline_ids = \App\Models\UserRelationships::withRecursiveQueryConstraint(function ($q) use ($date) {
                $q->whereDate('user_relationships.created_at', '<=', $date  )
                    ->where(function ($qq) use ($date) {
                        $qq->whereNull('user_relationships.superseded_at')
                            ->orWhereDate('user_relationships.superseded_at', '>', $date  );
                    });
            }, function () use ($date) {
                return \App\Models\UserRelationships::where('user_id', 437)
                                        ->latest()
                                        ->first()
                                        ->ancestors()
                                        ->where('created_at', '<=', $date ) // time contraint. We do include the time of the sale. So the customer of this sale does count
                                        ->where('superseded_at', '>', $date )
                                        ->pluck('user_id')
                                        ->toArray();
            });
        echo implode('<br>', $upline_ids );

        $queries = Illuminate\Support\Facades\DB::getQueryLog();
        dd ( end( $queries ));

    });
```

then I can get rid of the superseded relationship. BUT it still looks up the ancestors of the superseded relation:

```
234
24
25
24
array:3 [▼ // routes/web.php:69
  "query" => "with recursive `laravel_cte` as ((select *, -1 as `depth`, cast(`user_id` as char(65535)) as `path` from `user_relationships` where `user_relationships`.`user_id` = ?) union all (select `user_relationships`.*, `depth` - 1 as `depth`, concat(`path`, ?, `user_relationships`.`user_id`) from `user_relationships` inner join `laravel_cte` on `laravel_cte`.`sponsor_id` = `user_relationships`.`user_id` where date(`user_relationships`.`created_at`) <= ? and (`user_relationships`.`superseded_at` is null or date(`user_relationships`.`superseded_at`) > ?))) select `user_id` from `laravel_cte` where `created_at` <= ? and `superseded_at` > ? ◀"
  "bindings" => array:6 [▼
    0 => 234
    1 => "."
    2 => "2022-11-03"
    3 => "2022-11-03"
    4 => Carbon\Carbon @1667476800 {[#1584 ▶](http://sail.localhost/testrelationships#sf-dump-487463859-ref21584)}
    5 => Carbon\Carbon @1667476800 {[#1584 ▶](http://sail.localhost/testrelationships#sf-dump-487463859-ref21584)}
  ]
  "time" => 8.4
]
```

**Comment by staudenmeir** [comment]
> Now, I have to make these relationships time dependent. Meaning that there might be many parents for each user, but I'll be normally considering only the LATEST record, except when a date is provided.

Now you want to get the latest descendant (or ancestor, respectively)?

**Comment by valentinomariotto** [comment]
Hi! I want to be able to retrieve a tree of descendants or a line of ancestors at any point in time.
If no date is provided, then the resulting tree should be the current one (which will use the latest ancestor).

It seems to me that this is the part that's not working:
`select `user_id` from `laravel_cte` where `created_at` <= ? and `superseded_at``

**Comment by staudenmeir** [comment]
With the sample data from https://github.com/staudenmeir/laravel-adjacency-list/issues/144#issuecomment-1326552305, what result should the query give you for each of the users (and why)?

**Comment by valentinomariotto** [comment]
ancestors() query from user 437 should produce:
- 234, 25, 24 at any date past 2022-10-21 12:00:00
- 234, 24 at any date prior to 2022-10-21 12:00:00
because the relationship 24>234 had been superseded by the relationship 25>234 on that date.


descendants() query from user 25 should produce:
- 437 at any date prior to 2022-10-21 12:00:00
- 234, 437 at any date past 2022-10-21 12:00:00
because 234 was not a descendant of 25 prior to that date

descendants() query from user 24 should produce:
- 25, 234, 437 at any date
because 24 is at the highest position in this tree

**Comment by staudenmeir** [comment]
> descendants() query from user 25 should produce:
> 437 at any date prior to 2022-10-21 12:00:00

You don't want to stop when a relationship like `25 -> 234` is excluded but continue traversing down the tree? So unlike how [`withRecursiveQueryConstraint()`](https://github.com/staudenmeir/laravel-adjacency-list#recursive-query-constraints) behaves?

**Comment by valentinomariotto** [comment]
pardon me Jonas, I'm not sure I understand this question. 

if you consider the ancestors() of 437 at the present date, then the recursive query should:
```
- who's 437's parent? 
 > 234
  - who's 234's parent?
   >24: **but this relation expired on 2022-10-21, so ignore it and do not look up the parents**
   >25
   - who's 25's parent?
    >24
     - who's 24's parent?
      > 1 root. stop the recursion
```
Resulting int: 234, 25, 24.

similarly, if you consider the descendants() of 24 at the present date, then we have this logic:
```
- who are 24's children?
 > 25
  - who are 25's children?
   >234, but only after 2022-10-21
   - who are 234's children?
    > 437
    > and others
   >and others
 **> 234, but only until 2022-10-21, so ignore this and do not expand the children tree**
 > and others
```
resulting in: 25, 234, 437 ... and others

thank you for helping out!

**Comment by valentinomariotto** [comment]
ok, I found that **this works fine for the descendants**:

```
 $upline_ids = \App\Models\UserRelationships::withRecursiveQueryConstraint(function ($q) use ($date) {
            $q->where('user_relationships.superseded_at', '>', $date )
            ->where('user_relationships.created_at', '<=', $date );
        }, function () use ($date) {
            return \App\Models\UserRelationships::where('user_id', 25)
                                    ->latest()
                                    ->first()
                                    ->descendants()
                                    ->where('superseded_at', '>', $date )
                                    ->where('created_at', '<=', $date )
                                    ->pluck('user_id')
                                    ->toArray();
        });
```

I need to apply the condition twice, because the RecursiveQueryConstraint does not apply to the first iteration. 


On the other hand, **it still does not work for the ancestors.** Same problem as already described.

**Comment by valentinomariotto** [comment]
one last thing.. is it possible to apply the RecursiveQueryConstraint by default through some sort of scope, rather than writing it every time? Thank you

**Comment by valentinomariotto** [comment]
actually no, this still fails for immediate parents. It exapands the ancestor line when it should not.

it works if the parent is not immediate.

for example:

```
|user_id|sponsor_id|matrix_level|created_at         |updated_at         |superseded_at     |
|-------|----------|------------|-------------------|-------------------|-------------------|
|24     |1         |1           |2022-01-31 17:08:18|2022-11-24 09:21:13|2037-12-31 00:00:00|
|25     |24        |2           |2022-01-31 17:32:32|2022-11-24 09:21:13|2037-12-31 00:00:00|
|234    |24        |3           |2022-04-13 13:49:50|2022-11-24 09:21:13|2022-10-21 12:00:00|
|234    |25        |3           |2022-10-21 12:00:00|2022-11-24 09:21:13|2037-12-31 00:00:00|
|437    |234       |4           |2022-06-23 16:24:12|2022-11-24 09:21:14|2037-12-31 00:00:00|
|511    |437       |5           |2022-11-03 18:14:56|2022-11-03 18:14:56|2037-12-31 00:00:00|
```
- get ancestors of 511 at present time: works, producing 437, 234, 25, 24, 1
- get ancestors of 437 at present time: fails, producing 234, 24, 25, 1, 24, 1

**Comment by staudenmeir** [comment]
The issue looks very much like #117. Can you try the suggested workaround https://github.com/staudenmeir/laravel-adjacency-list/issues/117#issuecomment-1174128600?
